server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -677,8 +677,19 @@ public void updateShardState(
                                 // this Shard's engine was read only, we need to update its engine before restoring local history from xlog.
                                 assert newRouting.primary() && currentRouting.primary() == false;
                                 resetEngineToGlobalCheckpoint();
+                                // It is possible an engine can open with a SegmentInfos on a higher gen but the reader does not refresh to
+                                // trigger our refresh listener.
+                                // Force update the checkpoint post engine reset.
+                                updateReplicationCheckpoint();
                             }
+
                             replicationTracker.activatePrimaryMode(getLocalCheckpoint());
+                            if (indexSettings.isSegRepEnabled()) {
```

Comment:
Does this needs to be executed on node-node i.e. indexSettings.isSegRepLocalEnabled()?